include/effolkronium/random.hpp

```diff
@@ -642,6 +651,46 @@ namespace effolkronium {
             return dist( engine_instance( ) );
         }
 
+        /**
+        * \brief Return a random iterator from given map container by
+        *        utilizing the values of the map container as weights
+        *        for weighted random number generation
+        * \param Key The Key type for this version of 'get' method
+        *        Type should be '(THIS_TYPE)::common' struct
+        * \param map_container A container that has mapped_type,
+        *        value_type and key_type defined
+        * \return A random iterator from given map container based
+        *         on the result of the weighted random number generation
+        * \note return the end iterator if the iterator is empty
+        */
+        template<
+            typename Key,
+            class MapContainer
+        >
+        static auto get(const MapContainer& map_container) -> typename std::enable_if<
+            details::is_map<MapContainer>::value &&
+            details::is_iterator<decltype(std::begin(map_container))>::value &&
+            details::is_supported_number<typename MapContainer::mapped_type>::value &&
+            std::is_same<Key, common>::value,
+        decltype(std::begin(map_container))>::type {
+            using MappedType = typename MapContainer::mapped_type;
+            using PairType = typename MapContainer::value_type;
+            using IteratorType = decltype(std::begin(map_container));
+            
+            MappedType total_weight = std::accumulate(std::begin(map_container), std::end(map_container), 0, [](MappedType val, const PairType& p){
+                return val + p.second;
+            });
+            MappedType random_weight = get(0, total_weight - 1);
+            MappedType sum = 0;
+
+            for(IteratorType it = std::begin(map_container); it != std::end(map_container); ++it)
```

Comment:
By iterating through map we're missing reason of using the map instead of `vector<pair<Key, Value>>`

And for unordered map the order of weights will be different and result of this function will be different

I'm not expert in this algorithm, can it be done with map::upper_bound or map::lower_bound function ?
https://en.cppreference.com/w/cpp/container/map/upper_bound
for logarithmic complexity ?

Reply:
The upper_bound and lower_bound functions operate on the keys of the map so, it is not suitable for this case. And even if it operated on values, we had to store total weight until that object instead of the weight of that object as the value.

Also, the "ordering" of std::unordered_map should not affect the result. Because the probability of each item getting picked is directly tied to its weight.

Reply:
Fine, can you code a simple test case for this inside `test/random_test.cpp` file ? 

I can do it by myself but later next week

Reply:
During testing, I found out the algorithm fails for floats and doubles as values. I am also fixing that. Also, do you have a specific test case that you want me to code.

Reply:
I have nothing special, you can try your best

Reply:
All done. But I have one question, should I check whether if there is a negative weight using `assert`?

Reply:
> All done. But I have one question, should I check whether if there is a negative weight using `assert`?

It's much better to not allow users to use negative values at all by disallowing it with extra condition for enable_if

Something like this:
`!std::is_signed<MapContainer::value_type >::value`
And compilation should fail with signed map value types, so users required to use only unsigned types

Reply:
For real numbers (float, double, long double) assert is ok

Reply:
But this completely disallows users to use reals as weights.

Reply:
> 

!std::is_signed<MapContainer::value_type>::value || is_uniform_real<MapContainer::value_type>::value
...
assert(is_uniform_real<MapContainer::value_type>::value && map_has_all_positive_real_numbers())

Reply:
Done!